runtime/src/main/java/org/corfudb/runtime/object/DiskBackedSMRSnapshot.java

```diff
@@ -0,0 +1,107 @@
+package org.corfudb.runtime.object;
+
+import lombok.NonNull;
+import org.corfudb.runtime.collections.RocksDbEntryIterator;
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.util.serializer.ISerializer;
+import org.rocksdb.OptimisticTransactionDB;
+import org.rocksdb.ReadOptions;
+import org.rocksdb.Snapshot;
+import org.rocksdb.Transaction;
+import org.rocksdb.WriteOptions;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.concurrent.locks.StampedLock;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static org.corfudb.runtime.collections.RocksDbEntryIterator.LOAD_VALUES;
+
+public class DiskBackedSMRSnapshot<S extends SnapshotGenerator<S>> implements SMRSnapshot<S> {
+
+    private final VersionedObjectIdentifier version;
+    private final StampedLock lock = new StampedLock();
+    private final OptimisticTransactionDB rocksDb;
+    private final Snapshot snapshot;
+    private final ViewGenerator<S> viewGenerator;
+    private final ReadOptions readOptions;
+    private final WriteOptions writeOptions;
+    private final ColumnFamilyRegistry columnFamilyRegistry;
+
+    // A set of iterators associated with this snapshot.
+    private final Set<RocksDbEntryIterator<?, ?>> set;
+
+    public DiskBackedSMRSnapshot(@NonNull OptimisticTransactionDB rocksDb,
+                                 @NonNull WriteOptions writeOptions,
+                                 @NonNull VersionedObjectIdentifier version,
+                                 @NonNull ViewGenerator<S> viewGenerator,
+                                 @NonNull ColumnFamilyRegistry columnFamilyRegistry) {
+        this.rocksDb = rocksDb;
+        this.writeOptions = writeOptions;
+        this.viewGenerator = viewGenerator;
+        this.snapshot = rocksDb.getSnapshot();
```

Comment:
How do we make sure that this RockDB snapshot is equivalent to the table's view at this provided 'version'?

Reply:
Good catch! This is only used for debugging, but the version passed to it should not be based on `materializedUpTo`, but `globalAddress` instead. Here is the relevant block:


```
            if (globalAddress >= materializedUpTo) {
                final VersionedObjectIdentifier previousId = new VersionedObjectIdentifier(getID(), materializedUpTo);
                final VersionedObjectIdentifier currentId = new VersionedObjectIdentifier(getID(), globalAddress);
                currentObject.generateIntermediarySnapshot(currentId, objectOpenOption)
                        .ifPresent(newSnapshot -> {
                            final SMRSnapshot<S> previousSnapshot = setCurrentSnapshot(newSnapshot);
                            if (globalAddress == materializedUpTo) {
                                // The checkpoint sync will contain multiple SMR
                                // entries with the same address. Only cache the
                                // latest one, and release all the previous ones.
                                previousSnapshot.release();
                            }
                            mvoCache.put(previousId, previousSnapshot);
                        });
            }
````